// @wdio/browserstack-service/build/testOps/featureStats.js
import { BStackLogger } from '../bstackLogger.js';
import { isObjectEmpty } from '../util.js';
class FeatureStats {
    triggeredCount = 0;
    sentCount = 0;
    failedCount = 0;
    groups = {};
    mark(status, groupId) {
        switch (status) {
            case 'triggered':
                this.triggered(groupId);
                break;
            case 'success':
            case 'sent':
                this.sent(groupId);
                break;
            case 'failed':
                this.failed(groupId);
                break;
            default:
                BStackLogger.debug('Request to mark usage for unknown status - ' + status);
                break;
        }
    }
    triggered(groupId) {
        this.triggeredCount += 1;
        if (groupId) {
            this.createGroup(groupId).triggered();
        }
    }
    sent(groupId) {
        this.sentCount += 1;
        if (groupId) {
            this.createGroup(groupId).sent();
        }
    }
    failed(groupId) {
        this.failedCount += 1;
        if (groupId) {
            this.createGroup(groupId).failed();
        }
    }
    success(groupId) {
        this.sent(groupId);
    }
    createGroup(groupId) {
        if (!this.groups[groupId]) {
            this.groups[groupId] = new FeatureStats();
        }
        return this.groups[groupId];
    }
    getTriggeredCount() {
        return this.triggeredCount;
    }
    getSentCount() {
        return this.sentCount;
    }
    getFailedCount() {
        return this.failedCount;
    }
    getUsageForGroup(groupId) {
        return this.groups[groupId] || new FeatureStats();
    }
    getOverview() {
        return { triggeredCount: this.triggeredCount, sentCount: this.sentCount, failedCount: this.failedCount };
    }
    getGroups() {
        return this.groups;
    }
    add(featureStats) {
        this.triggeredCount += featureStats.getTriggeredCount();
        this.sentCount += featureStats.getSentCount();
        this.failedCount += featureStats.getFailedCount();
        Object.entries(featureStats.getGroups()).forEach(([groupId, group]) => {
            this.createGroup(groupId).add(group);
        });
    }
    // omitGroups: true/false -> Include groups or not
    // onlyGroups: true/false -> data includes only groups
    // nestedGroups: true/false -> groups will be nested in groups if true
    toJSON(config = {}) {
        const overviewData = !config.onlyGroups ? {
            triggeredCount: this.triggeredCount,
            sentCount: this.sentCount,
            failedCount: this.failedCount
        } : {};
        const groupsData = {};
        if (!config.omitGroups) {
            Object.entries(this.groups).forEach(([groupId, group]) => {
                groupsData[groupId] = group.toJSON(); // Currently Nested groups are only overviews
            });
        }
        const group = config.nestedGroups ? { groups: groupsData } : groupsData;
        return {
            ...overviewData,
            ...group
        };
    }
    static fromJSON(json) {
        const stats = new FeatureStats();
        if (!json || isObjectEmpty(json)) {
            return stats;
        }
        stats.triggeredCount = json.triggeredCount;
        stats.sentCount = json.sentCount;
        stats.failedCount = json.failedCount;
        if (!json.groups) {
            return stats;
        }
        Object.entries(json.groups).forEach(([groupId, group]) => {
            stats.groups[groupId] = FeatureStats.fromJSON(group);
        });
        return stats;
    }
}
export default FeatureStats;
