// @wdio/browserstack-service/build/config.js
import TestOpsConfig from './testOps/testOpsConfig.js';
import { isUndefined } from './util.js';
class BrowserStackConfig {
    static getInstance(options, config) {
        if (!this._instance && options && config) {
            this._instance = new BrowserStackConfig(options, config);
        }
        return this._instance;
    }
    userName;
    accessKey;
    framework;
    buildName;
    buildIdentifier;
    testObservability;
    percy;
    accessibility;
    app;
    static _instance;
    appAutomate;
    automate;
    funnelDataSent = false;
    constructor(options, config) {
        this.framework = config.framework;
        this.userName = config.user;
        this.accessKey = config.key;
        this.testObservability = new TestOpsConfig(options.testObservability !== false, !isUndefined(options.testObservability));
        this.percy = options.percy || false;
        this.accessibility = options.accessibility || false;
        this.app = options.app;
        this.appAutomate = !isUndefined(options.app);
        this.automate = !this.appAutomate;
        this.buildIdentifier = options.buildIdentifier;
    }
    sentFunnelData() {
        this.funnelDataSent = true;
    }
}
export default BrowserStackConfig;
