// @wdio/browserstack-service/build/Percy/PercyLogger.js
import path from 'node:path';
import fs from 'node:fs';
import chalk from 'chalk';
import logger from '@wdio/logger';
import { PERCY_LOGS_FILE } from '../constants.js';
import { COLORS } from '../util.js';
const log = logger('@wdio/browserstack-service');
export class PercyLogger {
    static logFilePath = path.join(process.cwd(), PERCY_LOGS_FILE);
    static logFolderPath = path.join(process.cwd(), 'logs');
    static logFileStream;
    static logToFile(logMessage, logLevel) {
        try {
            if (!this.logFileStream) {
                if (!fs.existsSync(this.logFolderPath)) {
                    fs.mkdirSync(this.logFolderPath);
                }
                this.logFileStream = fs.createWriteStream(this.logFilePath, { flags: 'a' });
            }
            if (this.logFileStream && this.logFileStream.writable) {
                this.logFileStream.write(this.formatLog(logMessage, logLevel));
            }
        }
        catch (error) {
            log.debug(`Failed to log to file. Error ${error}`);
        }
    }
    static formatLog(logMessage, level) {
        return `${chalk.gray(new Date().toISOString())} ${chalk[COLORS[level]](level.toUpperCase())} ${chalk.whiteBright('@wdio/browserstack-service')} ${logMessage}\n`;
    }
    static info(message) {
        this.logToFile(message, 'info');
        log.info(message);
    }
    static error(message) {
        this.logToFile(message, 'error');
        log.error(message);
    }
    static debug(message, param) {
        this.logToFile(message, 'debug');
        if (param) {
            log.debug(message, param);
        }
        else {
            log.debug(message);
        }
    }
    static warn(message) {
        this.logToFile(message, 'warn');
        log.warn(message);
    }
    static trace(message) {
        this.logToFile(message, 'trace');
        log.trace(message);
    }
    static clearLogger() {
        if (this.logFileStream) {
            this.logFileStream.end();
        }
        this.logFileStream = null;
    }
    static clearLogFile() {
        if (fs.existsSync(this.logFilePath)) {
            fs.truncateSync(this.logFilePath);
        }
    }
}
